### ISSUE TITLE
Manual jail start failed with UnknownJailException

### ISSUE THREAD

**Issue by CarstenGrohmann** [fail2ban/fail2ban#1114]
Hi,

I've installed Fail2ban 0.9.2 and configured an sshd jail. The jail will be successfully started during the automatic startup of Fail2ban, but it failed on manual start:
```
# fail2ban-client start sshd
ERROR  NOK: ('sshd',)
Sorry but the jail 'sshd' does not exist
```

What's the reason of this error?

Steps to reproduce:

1. Install Fail2ban 0.9.2 from EPEL repository on RHEL7

2. Create configuration files

     ```
# cat /etc/fail2ban/jail.d/010_init.conf
[DEFAULT]
ignoreip = 127.0.0.1 172.250.1.2
backend = systemd

# cat /etc/fail2ban/jail.d/020_ssh.conf
[sshd]
enabled = true
port    = ssh
filter  = sshd
     ```

3. Start Fail2ban and check status:

    ```
# fail2ban-client status
Status
|- Number of jail:      1
`- Jail list:   sshd
    ```

4. Shutdown sshd jail:

    ```
# fail2ban-client stop sshd
Jail stopped

# fail2ban-client status
Status
|- Number of jail:      0
`- Jail list:
    ```

5. Restart sshd jail

    ```
# fail2ban-client start sshd
ERROR  NOK: ('sshd',)
Sorry but the jail 'sshd' does not exist
    ```

6. Content of */var/log/fail2ban.log* (loglevel DEBUG):

    ```
# tail -n 30 /var/log/fail2ban.log
iptables -A f2b-sshd -j RETURN
iptables -I INPUT -p tcp -m multiport --dports ssh -j f2b-sshd -- stdout: ''
2015-07-13 11:42:34,065 fail2ban.action         [31764]: DEBUG   iptables -N f2b-sshd
iptables -A f2b-sshd -j RETURN
iptables -I INPUT -p tcp -m multiport --dports ssh -j f2b-sshd -- stderr: ''
2015-07-13 11:42:34,065 fail2ban.action         [31764]: DEBUG   iptables -N f2b-sshd
iptables -A f2b-sshd -j RETURN
iptables -I INPUT -p tcp -m multiport --dports ssh -j f2b-sshd -- returned successfully
2015-07-13 11:42:48,369 fail2ban.transmitter    [31764]: DEBUG   Command: ['status']
2015-07-13 11:42:53,769 fail2ban.transmitter    [31764]: DEBUG   Command: ['stop', 'sshd']
2015-07-13 11:42:53,769 fail2ban.server         [31764]: DEBUG   Stopping jail sshd
2015-07-13 11:42:53,997 fail2ban.filtersystemd  [31764]: DEBUG   sshd filter terminated
2015-07-13 11:42:54,086 fail2ban.actions        [31764]: DEBUG   Flush ban list
2015-07-13 11:42:54,087 fail2ban.action         [31764]: DEBUG   iptables -D INPUT -p tcp -m multiport --dports ssh -j f2b-sshd
iptables -F f2b-sshd
iptables -X f2b-sshd
2015-07-13 11:42:54,191 fail2ban.action         [31764]: DEBUG   iptables -D INPUT -p tcp -m multiport --dports ssh -j f2b-sshd
iptables -F f2b-sshd
iptables -X f2b-sshd -- stdout: ''
2015-07-13 11:42:54,191 fail2ban.action         [31764]: DEBUG   iptables -D INPUT -p tcp -m multiport --dports ssh -j f2b-sshd
iptables -F f2b-sshd
iptables -X f2b-sshd -- stderr: ''
2015-07-13 11:42:54,191 fail2ban.action         [31764]: DEBUG   iptables -D INPUT -p tcp -m multiport --dports ssh -j f2b-sshd
iptables -F f2b-sshd
iptables -X f2b-sshd -- returned successfully
2015-07-13 11:42:54,191 fail2ban.actions        [31764]: DEBUG   sshd: action terminated
2015-07-13 11:42:54,191 fail2ban.jail           [31764]: INFO    Jail 'sshd' stopped
2015-07-13 11:42:56,705 fail2ban.transmitter    [31764]: DEBUG   Command: ['status']
2015-07-13 11:43:34,160 fail2ban.transmitter    [31764]: DEBUG   Command: ['start', 'sshd']
2015-07-13 11:43:34,160 fail2ban.transmitter    [31764]: WARNING Command ['start', 'sshd'] has failed. Received UnknownJailException('sshd',)
    ```

Regards,
Carsten

**Comment by sebres** [comment]
Currently stopping the jail removes it completely from current running fail2ban-server, so you should to add it again, but also set again all the parameters manual also (what would be not so easy).
```
fail2ban-client add sshd <BACKEND>
fail2ban-client set sshd addignoreregex <REGEX>
fail2ban-client set sshd delignoreregex <INDEX>
...
fail2ban-client start sshd
```
Or just use `reload` instead of.

I will make a review what we can do, to transfer automatically all parameters single jail only.

**Comment by sebres** [comment]
Because of meanwhile very complex algorithm to pass parameters to jail, like the filter/action includes, it internal substitution (recursively substitution of substited entries) etc, we should implement a command like `reload <jail>`, that can add a single jail with all it parameters (to start it again with `start <jail>`) or if already active simply reload it.

**Comment by yarikoptic** [comment]
On Thu, 16 Jul 2015, Serg G. Brester wrote:

>    Because of meanwhile very complex algorithm to pass parameters to jail,
>    like the filter/action includes, it internal substitution (recursively
>    substitution of substited entries) etc, we should implement a command like
>    reload <jail>, that can add a single jail with all it parameters (to start
>    it again with start <jail>) or if already active simply reload it.

don't we have it already?

       reload <JAIL>
              reloads the jail <JAIL>

**Comment by sebres** [comment]
Indeed, but I'm not sure that will work if the jail was stopped...

**Comment by sebres** [comment]
I have tested it: seems to be work, but If a jail exists it will be reloaded without any answer, if it was already stopped - reload but returns NOK.
```
$ f2b status
Status
|- Number of jail:      2
`- Jail list:   pam-generic, test-sb
$ f2b reload test-sb
??? nothing as answer ???
$ f2b status
Status
|- Number of jail:      2
`- Jail list:   pam-generic, test-sb
$ f2b stop test-sb
Jail stopped
$ f2b reload test-sb
ERROR  NOK: ('test-sb',)
$ f2b 
Status
|- Number of jail:      2
`- Jail list:   pam-generic, test-sb
```
Here is a log excerpt (one warning present):
```
2015-07-17 13:12:15,301 fail2ban.transmitter    [22776]: DEBUG   Command: ['ping']
2015-07-17 13:12:15,316 fail2ban.transmitter    [22776]: DEBUG   Command: ['stop', 'test-sb']
2015-07-17 13:12:15,317 fail2ban.server         [22776]: DEBUG   Stopping jail test-sb
2015-07-17 13:12:15,317 fail2ban.transmitter    [22776]: WARNING Command ['stop', 'test-sb'] has failed. Received UnknownJailException('test-sb',)
2015-07-17 13:12:15,317 fail2ban.transmitter    [22776]: DEBUG   Command: ['set', 'loglevel', 'DEBUG']
2015-07-17 13:12:15,318 fail2ban.transmitter    [22776]: DEBUG   Command: ['set', 'logtarget', '/var/log/fail2ban.log']
2015-07-17 13:12:15,318 fail2ban.server         [22776]: INFO    Changed logging target to /var/log/fail2ban.log for Fail2ban v0.9.2.dev0
2015-07-17 13:12:15,319 fail2ban.transmitter    [22776]: DEBUG   Command: ['set', 'syslogsocket', 'auto']
2015-07-17 13:12:15,319 fail2ban.transmitter    [22776]: DEBUG   Command: ['set', 'dbfile', '/var/run/fail2ban.sqlite3']
2015-07-17 13:12:15,319 fail2ban.transmitter    [22776]: DEBUG   Command: ['set', 'dbpurgeage', '86400']
2015-07-17 13:12:15,320 fail2ban.transmitter    [22776]: DEBUG   Command: ['add', 'test-sb', 'pyinotify']
2015-07-17 13:12:15,320 fail2ban.jail           [22776]: INFO    Creating new jail 'test-sb'
...
```

**Comment by sebres** [comment]
yep, NOK "throws" command stop (only if level is DEBUG), so this issue can be closed.

**Comment by osiyuk** [comment]
The problem still persists

**Comment by sebres** [comment]
> The problem still persists

Use `reload <jail>` instead of `start <jail>`. Command `start` requires add jail and set all the config entries manually.

**Comment by snn1024a** [comment]
I know that's an old topic but I've just got the same problem
restarting the client using : 
$fail2ban-client stop
and
$fail2ban-client start
worked just fine

 $fail2ban-client status
Status
|- Number of jail:	1
`- Jail list:	sshd

$iptables -n -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
f2b-sshd   tcp  --  0.0.0.0/0            0.0.0.0/0            multiport dports 22
Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain f2b-sshd (1 references)
target     prot opt source               destination
REJECT     all  --  123.207.168.222      0.0.0.0/0            reject-with icmp-port-unreachable
REJECT     all  --  206.189.146.13        0.0.0.0/0            reject-with icmp-port-unreachable
REJECT     all  --  178.62.235.23         0.0.0.0/0            reject-with icmp-port-unreachable
REJECT     all  --  173.212.254.167      0.0.0.0/0            reject-with icmp-port-unreachable
REJECT     all  --  118.24.146.83          0.0.0.0/0            reject-with icmp-port-unreachable
REJECT     all  --  218.92.0.197            0.0.0.0/0            reject-with icmp-port-unreachable
RETURN     all  --  0.0.0.0/0                 0.0.0.0/0

**Comment by sebres** [comment]
@snn1024a fail2ban version? 

The reload on jail should be fixed in 0.10...
If it was a new jail, before 0.10 one need firstly to add it and then supply all expected parameters before one would enable it (see a dump of that using `fail2ban-client -d | grep <jail>`).

Or as mentioned in comment https://github.com/fail2ban/fail2ban/issues/1114#issuecomment-306402120:
> Use `reload <jail>` instead of `start <jail>`. Command `start` requires add jail and set all the config entries manually.

Of course you can reload would fail2ban configuration (firstly introduced in 0.10, as fast reload without unbanning etc like using restart), so just enable jail in your `jail.local` config and  do:
`?sudo? fail2ban-client reload`.
